### PR TITLE
Add support for collections

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️/❌

### CODE REVIEW

File: docs/guide/en/typecasting.md [hunk 1]
```diff
@@ -20,9 +20,10 @@ $lock = $hydrator->create(Lock::class, ['name' => 'The lock', 'isLocked' => 1]);
...
 
 ```php
+use Yiisoft\Hydrator\Hydrator;
 use Yiisoft\Hydrator\TypeCaster\CompositeTypeCaster;
 use Yiisoft\Hydrator\TypeCaster\PhpNativeTypeCaster;
-use Yiisoft\Hydrator\TypeCaster\HydratorTypeCaster
+use Yiisoft\Hydrator\TypeCaster\HydratorTypeCaster;
 
 $typeCaster = new CompositeTypeCaster(
     new PhpNativeTypeCaster(),
```
Comment (arogachev) [inline docs/guide/en/typecasting.md:29]: Need to add the section about collections to the guide. Can be done in a separate PR.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 ## 1.3.0 under development
 
+- New #49: Add support for collections (@arogachev)
```
Comment (vjik) [inline CHANGELOG.md:5]: Add info about new attribute and changes in attribute handling classes

File: docs/guide/en/typecasting.md [hunk 3]
```diff
@@ -119,6 +120,8 @@ echo $post->getAuthor()->getNickName();
 
 ## Using attributes
 
+### Strings
```
Comment (vjik) [inline docs/guide/en/typecasting.md:123]: Let's use attribute names as headings. WDYT?

File: src/Attribute/Parameter/CollectionResolver.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            return Result::fail();
+        }
+
+        $resolvedValue = $context->getResolvedValue();
+        if (!is_iterable($resolvedValue)) {
+            return Result::fail();
+        }
+
+        $collection = [];
+        foreach ($context->getResolvedValue() as $item) {
```
Comment (vjik) [inline src/Attribute/Parameter/CollectionResolver.php:32]:
```suggestion
        foreach ($resolvedValue as $item) {
```

File: docs/guide/en/typecasting.md [hunk 5]
```diff
@@ -156,20 +179,94 @@ class Person
...
+$category = $hydrator->create(
+    PostCategory::class,
+    [
+        ['name' => 'Post 1'],
+        ['name' => 'Post 2', 'description' => 'Description for post 2'],
+    ],
+);
+```
+
+One-to-many relations can be nested and combined with one-to-one relations:
```
Comment (vjik) [inline docs/guide/en/typecasting.md:214]: Are you sure, that "One-to-many relations" is need? I think it's unnecessary...

File: src/AttributeHandling/ParameterAttributeResolveContext.php [hunk 6]
```diff
@@ -23,6 +24,7 @@ public function __construct(
         private ReflectionParameter|ReflectionProperty $parameter,
         private Result $resolveResult,
         private DataInterface $data,
+        private HydratorInterface $hydrator,
```
Comment (vjik) [inline src/AttributeHandling/ParameterAttributeResolveContext.php:27]:
```suggestion
        private ?HydratorInterface $hydrator = null,
```

And throw `LogicException` in `getHydrator()` when hydrator is not set. It's need for BC. We make hydrator required in next major version.

File: src/AttributeHandling/ParameterAttributesHandler.php [hunk 7]
```diff
@@ -60,7 +62,8 @@ public function handle(
                 );
             }
 
-            $context = new ParameterAttributeResolveContext($parameter, $resolveResult, $data);
+            $this->requireHydrator();
```
Comment (vjik) [inline src/AttributeHandling/ParameterAttributesHandler.php:65]: Move this call out of foreach.

File: src/AttributeHandling/ParameterAttributesHandler.php [hunk 8]
```diff
@@ -70,4 +73,18 @@ public function handle(
...
     }
+
+    /**
+     * Ensure that validator is set in parameter attributes handler.
+     *
+     * @psalm-assert HydratorInterface $this->hydrator
+     *
+     * @throws RuntimeException If hydrator is not set in parameter attributes handler.
+     */
+    private function requireHydrator(): void
```
Comment (vjik) [inline src/AttributeHandling/ParameterAttributesHandler.php:84]: This method call once, may be use code directly?

File: src/AttributeHandling/ParameterAttributesHandler.php [hunk 9]
```diff
@@ -22,6 +23,7 @@ final class ParameterAttributesHandler
 {
     public function __construct(
         private AttributeResolverFactoryInterface $attributeResolverFactory,
+        private ?HydratorInterface $hydrator = null,
```
Comment (vjik) [inline src/AttributeHandling/ParameterAttributesHandler.php:26]: Add also method `withHydrator()`.

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/AttributeHandling/ParameterAttributesHandler.php [hunk 10]
```diff
@@ -70,4 +77,11 @@ public function handle(
 
         return $resolveResult;
     }
+
+    public function withHydrator(HydratorInterface $hydrator): static
```
Comment (vjik) [inline src/AttributeHandling/ParameterAttributesHandler.php:81]:
```suggestion
    public function withHydrator(HydratorInterface $hydrator): self
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -2,6 +2,10 @@
 
 ## 1.3.0 under development
 
+- New #49: Add support for collections:
+  - Add `Collection` attribute with handler for using with parameters;
+  - Add hydrator dependency to `ParameterAttributeResolveContext` and `ParameterAttributesHandler`.                         
+  (@arogachev)
```
Comment (vjik) [inline CHANGELOG.md:8]:
```suggestion
- New #49: Add `Collection` PHP attribute (@arogachev)
- New #49: Add hydrator dependency and `withHydrator()` method to `ParameterAttributesHandler` (@arogachev)
- New #49: Add hydrator dependency and `getHydrator()` method to `ParameterAttributeResolveContext` (@arogachev)
```

File: tests/Attribute/Parameter/CollectionTest.php [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+
+        $hydrator->hydrate(
+            $object,
+            [
+                'posts' => [
+                    new stdClass(),
+                ],
+            ],
+        );
+        $this->assertEquals($object->getPosts(), [new stdClass()]);
```
Comment (vjik) [inline tests/Attribute/Parameter/CollectionTest.php:82]: I think in this case result should be empty array.

File: docs/guide/en/typecasting.md [hunk 13]
```diff
@@ -156,20 +181,94 @@ class Person
...
+$category = $hydrator->create(
+    PostCategory::class,
+    [
+        ['name' => 'Post 1'],
+        ['name' => 'Post 2', 'description' => 'Description for post 2'],
+    ],
+);
+```
+
+Collections can be nested and combined one-to-one relations:
```
Comment (vjik) [inline docs/guide/en/typecasting.md:216]: But is it one-to-many relation?

**Review by vjik — CHANGES_REQUESTED** [review]

**Review by vjik — APPROVED** [review]

File: docs/guide/en/typecasting.md [hunk 14]
```diff
@@ -139,6 +142,28 @@ $money = $hydrator->create(Money::class, [
 ]);
 ```
 
+### `Trim` / `LeftTrim` / `RightTrim`
+
+To strip whitespace (or other characters) from the beginning and/or end of a resolved string value, you can use `Trim`,
+`LeftTrim` or `RightTrim` attributes:
```
Comment (samdark) [inline docs/guide/en/typecasting.md:148]: Is it better to name these `TrimLeft`/`TrimRight`?

**Review by samdark — APPROVED** [review]

File: docs/guide/en/typecasting.md (hunk 14 above)
Comment (arogachev) [reply docs/guide/en/typecasting.md:148]: Thanks for your suggestion. I've created separate issue - #91, because it's out of scope of current PR.

**Pull request merged**